src/linear_algebra/gram_schmidt_ortho.lean

```diff
@@ -0,0 +1,136 @@
+/-
+Copyright (c) 2022 Jiale Miao. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jiale Miao, Kevin Buzzard
+-/
+
+import tactic.basic
+import algebra.big_operators.basic
+import analysis.inner_product_space.basic
+import analysis.normed_space.is_R_or_C
+
+/-!
+# Gram-Schmidt Orthogonalization and Orthonormalization
+
+In this file we introduce Gram-Schmidt Orthogonalization and Orthonormalization
+
+## Main results
+
+- `proj` : projection between two vectors in the inner product space
+- `GS`   : Gram-Schmidt Process
+- `GS_Orthogonal` : the proof that "GS" produces an orthogonal system of vectors
+- `GS_unit` : Normalized "Gram-Schmidt" (i.e each vector in this system has unit length)
+- `GS_Orthornormal` : the proof that "GS_unit" produces an orthornormal system of vectors
+-/
+
+open_locale big_operators
+
+variables (𝕜 : Type*) (E : Type*) [is_R_or_C 𝕜] [inner_product_space 𝕜 E]
+
+/-- projection in the inner product space -/
+def proj (u v : E) : E := ((inner u v) / (inner u u) : 𝕜) • u
+
+/-- Definition of Gram-Schmidt Process -/
+def GS (f : ℕ → E) : ℕ → E
+| n := f n - ∑ i in finset.range(n),
+  if h1 : i < n then proj 𝕜 E (GS i) (f n) else f 37
+
+/-- This helps us to get rid of 'ite' in the definition of GS -/
+@[simp] lemma GS_n_1 (f : ℕ → E) (n : ℕ) :
+GS 𝕜 E f (n + 1) = f (n + 1) - ∑ i in finset.range(n + 1), proj 𝕜 E (GS 𝕜 E f i) (f (n + 1)) :=
+begin
+  rw [GS, sub_right_inj],
+  apply finset.sum_congr rfl,
+  intros x hx,
+  rw finset.mem_range at hx,
+  rw if_pos hx,
+end
+
+/-- inner product defined as add_monoid_hom -/
+def inner_product_hom (v : E) [inner_product_space 𝕜 E] : E →+ 𝕜 :=
+{ to_fun := inner v,
+  map_zero' := inner_zero_right,
+  map_add' := λ x y, inner_add_right }
+
+lemma inner_eq_inner_hom (v x : E) : inner v x = inner_product_hom 𝕜 E v x := rfl
+
+lemma inner_hom_eq_inner (v x : E) : inner_product_hom 𝕜 E v x = inner v x := rfl
+
+/-- # Gram-Schmidt Orthogonalisation -/
+theorem GS_Orthogonal (f : ℕ → E) (c : ℕ) :
+∀ (a b : ℕ), a < b → b ≤ c → (inner (GS 𝕜 E f a) (GS 𝕜 E f b) : 𝕜) = 0 :=
+begin
+  induction c with c hc,
+  { intros a b ha hb, simp at hb, simp [hb] at ha, contradiction },
+  { intros a b ha hb, rw nat.le_add_one_iff at hb, cases hb with hb₁ hb₂,
+    { specialize hc a b, simp [ha, hb₁] at hc, exact hc },
+    { simp [GS_n_1, hb₂, inner_sub_right],
+      rw [inner_eq_inner_hom, inner_eq_inner_hom],
+      rw map_sum (inner_product_hom 𝕜 E (GS 𝕜 E f a))
```

Comment:
We ought to have an `inner_sum` lemma somewhere - were you not able to find it?

Reply:
oh yes, I found that lemma, I have fixed it. Thank you very much!